cpp/include/raft/neighbors/detail/cagra/cagra_search.cuh

```diff
@@ -149,6 +149,71 @@ void search_main_core(
   }
 }
 
+template <class T,
+          class DatasetT,
+          class DatasetIdxT,
+          class InternalIdxT,
+          class DistanceT,
+          class CagraSampleFilterT>
+void lauch_vpq_search_main_core(
+  raft::resources const& res,
+  const vpq_dataset<DatasetT, DatasetIdxT>* dataset,
+  search_params params,
+  raft::device_matrix_view<const InternalIdxT, int64_t, row_major> graph,
+  raft::device_matrix_view<const T, int64_t, row_major> queries,
+  raft::device_matrix_view<InternalIdxT, int64_t, row_major> neighbors,
+  raft::device_matrix_view<DistanceT, int64_t, row_major> distances,
+  CagraSampleFilterT sample_filter)
+{
+  if (dataset->pq_bits() == 8) {
+    if (dataset->pq_len() == 2) {
+      using dataset_desc_t = cagra_q_dataset_descriptor_t<T,
+                                                          DatasetT,
+                                                          8 /*PQ bit*/,
+                                                          2 /* Subspace dimension*/,
+                                                          0,
+                                                          DistanceT,
+                                                          InternalIdxT,
+                                                          0>;
+      dataset_desc_t dataset_desc(dataset->data.data_handle(),
+                                  dataset->pq_dim(),
+                                  dataset->vq_n_centers(),
+                                  dataset->vq_code_book.data_handle(),
+                                  0,
+                                  dataset->pq_code_book.data_handle(),
+                                  0,
+                                  dataset->n_rows(),
+                                  dataset->dim());
+      search_main_core(
+        res, params, dataset_desc, graph, queries, neighbors, distances, sample_filter);
+    } else if (dataset->pq_len() == 4) {
+      using dataset_desc_t = cagra_q_dataset_descriptor_t<T,
+                                                          DatasetT,
+                                                          8 /*PQ bit*/,
+                                                          4 /* Subspace dimension*/,
+                                                          0,
+                                                          DistanceT,
+                                                          InternalIdxT,
+                                                          0>;
+      dataset_desc_t dataset_desc(dataset->data.data_handle(),
+                                  dataset->pq_dim(),
+                                  dataset->vq_n_centers(),
+                                  dataset->vq_code_book.data_handle(),
+                                  0,
+                                  dataset->pq_code_book.data_handle(),
+                                  0,
+                                  dataset->n_rows(),
+                                  dataset->dim());
```

Comment:
It's probably lost in merging, please check if the mapping from vpq_dataset to dataset_desc_t is correct. I believe it should be:
```
dataset_desc_t dataset_desc(vpq_dset->data.data_handle(),	
                                vpq_dset->encoded_row_length(),	
                                vpq_dset->pq_dim(),	
                                vpq_dset->vq_code_book.data_handle(),	
                                1.0f,	
                                vpq_dset->pq_code_book.data_handle(),	
                                1.0f,	
                                size_t(vpq_dset->n_rows()),	
                                vpq_dset->dim());
```

Reply:
Is it correct that the pq and vq scales are fixed as 1.f because only f16 is currently supported as a dataset type?

Reply:
Yes, as far as I understood it from the reference implementation, the scaling factors other than one are used only for (u)int8 types, which was also just an experiment there.